### PR TITLE
lease: deal with race when concurrent acquisition races with purge

### PR DESCRIPTION
**Pull request by ajwerner**
We had a race that used to be mitigated by the fact that we would always
mark dropped descriptors as taken offline. In general this saved us in the
case the descriptor was being dropped. If the descriptor was not being
dropped then generally schema changes would proceed because they would
avoid leasing and then subsequent updates to the descriptor would not
hit the race. This simple mechanism allows us to ensure that we drop
versions of a descriptor we know not to be the newest.

The behavior from https://github.com/cockroachdb/cockroach/pull/59375 exposed this bug.
Prior to that change, we would have marked the table as offline and thus always would have
released it.

Release justification: bug fixes and low-risk updates to new functionality

Release note (bug fix): Fixed a bug where sometimes when a descriptor is dropped and there is concurrent use, that descriptor's  lease is not released.

### CODE REVIEW

**Comment by ajwerner** [comment]
FYI @lucy-zhang this was the cause of the timeouts in the schemachange workload.

File: pkg/sql/catalog/lease/lease.go [hunk 1]
```diff
@@ -1726,6 +1738,7 @@ func (m *Manager) refreshLeases(
 					if err := evFunc(desc); err != nil {
 						log.Infof(ctx, "skipping update of %v due to knob: %v",
 							desc, err)
+						continue
```
Comment (postamar) [inline pkg/sql/catalog/lease/lease.go:1741]: I'm surprised that fixing this omission didn't reveal bugs in existing tests. Oh well.

File: pkg/sql/catalog/lease/lease.go [hunk 2]
```diff
@@ -1026,9 +1035,12 @@ func purgeOldVersions(
 		return nil
 	}
 	t.mu.Lock()
+	if t.mu.maxVersionSeen < minVersion {
+		t.mu.maxVersionSeen = minVersion
+	}
 	empty := len(t.mu.active.data) == 0 && t.mu.acquisitionsInProgress == 0
 	t.mu.Unlock()
-	if empty {
+	if empty && !takenOffline {
```
Comment (postamar) [inline pkg/sql/catalog/lease/lease.go:1043]: Nice catch.

**Review by postamar — APPROVED** [review]

**Comment by ajwerner** [comment]
TFTR!

bors r=postamar

**Pull request merged**